lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,2561 @@
+    begin = std::min(begin, other->begin);
+    end = std::max(end, other->end);
+  }
+
+  uint getBegin() { return begin; }
+  uint getEnd() { return end; }
+};
+
+class LifeTimeAnalysis {
+private:
+  StringRef name;
+  SmallVector<LifeTime *> lifetimes;
+  SetVector<PhysicalQID> frame;
+  // DenseMap<VirtualQID, PhysicalQID> allocMap;
+
+  /// Given a candidate lifetime, tries to find a qubit to reuse,
+  /// and otherwise allocates a new qubit
+  PhysicalQID allocatePhysical(LifeTime *lifetime) {
+    std::optional<size_t> best_reuse = std::nullopt;
+    std::optional<size_t> empty = std::nullopt;
+    uint best_distance = INT_MAX;
+
+    for (uint i = 0; i < lifetimes.size(); i++) {
+      if (!lifetimes[i]) {
+        empty = i;
+        continue;
+      }
+
+      auto other = lifetimes[i];
+      auto distance = lifetime->distance(other);
+      if (!lifetime->isOverlapping(other) && distance < best_distance) {
+        best_reuse = i;
+        best_distance = distance;
+      }
+    }
+
+    // Reuse a qubit based on its lifetime in the same scope
+    if (best_reuse) {
+      auto physical = best_reuse.value();
+      lifetimes[physical]->combine(lifetime);
+      delete lifetime;
+      return physical;
+    }
+
+    // Reuse a qubit without a lifetime (used in a different frame)
+    if (empty) {
+      auto physical = empty.value();
+      lifetimes[physical] = lifetime;
+      return physical;
+    }
+
+    // Fall back: allocate a new qubit
+    lifetimes.push_back(lifetime);
+    return lifetimes.size() - 1;
+  }
+
+public:
+  LifeTimeAnalysis(StringRef name) : name(name), lifetimes(), frame() {}
+
+  ~LifeTimeAnalysis() {
+    for (auto lifetime : lifetimes)
+      if (lifetime)
+        delete lifetime;
+  }
+
+  PhysicalQID allocatePhysical(VirtualQID qid, LifeTime *lifetime) {
+    auto phys = allocatePhysical(lifetime);
+    frame.insert(phys);
+    return phys;
+  }
+
+  SetVector<PhysicalQID> getAllocated() {
+    for (uint i = 0; i < lifetimes.size(); i++) {
+      if (lifetimes[i])
+        delete lifetimes[i];
+      lifetimes[i] = nullptr;
+    }
+    auto pqids = SetVector<PhysicalQID>(frame);
+    frame.clear();
+    return pqids;
+  }
+
+  void reallocatePhysical(PhysicalQID phys, LifeTime *lifetime) {
+    lifetimes[phys] = lifetime;
+  }
+
+  size_t getCount() { return lifetimes.size(); }
+
+  void print() {
+    llvm::outs() << "# qubits: " << getCount() << ", cycles: ";
+    for (size_t i = 0; i < lifetimes.size(); i++)
+      llvm::outs() << lifetimes[i]->getBegin() << " - "
+                   << lifetimes[i]->getEnd() << " ";
+    llvm::outs() << "\n";
+  }
+
+  StringRef getName() { return name; }
+};
+
+class DependencyGraph;
+
+class DependencyNode {
+  friend class DependencyGraph;
+  friend class OpDependencyNode;
+  friend class IfDependencyNode;
+  friend class ArgDependencyNode;
+  friend class RootDependencyNode;
+  friend class TerminatorDependencyNode;
+  friend class InitDependencyNode;
```

Comment:
This type is very gregarious.

Reply:
I've made the API for DependencyNodes a little cleaner, so hopefully most of those can be removed.

Reply:
Most of the ones have been removed, and the remaining ones have a decent reason for staying.